### PR TITLE
runner: add armv7a9-zynq7000-emu implementation and enable test-mkdir for imxrt targets

### PR DESCRIPTION
**Pull request by damianloew**
## Description

- runner: add armv7a9-zynq7000-emu implementation
- psh: disable asserting ownership in mkdir test and enable the test on dummyfs targets
- psh: test-mkdir: divide fname with all chars

## Motivation and Context

JIRA: CI-110
JIRA: CI-120

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: (imxrt eval kits, zynq7000 qemu).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [x] This PR needs additional PRs to work (https://github.com/phoenix-rtos/phoenix-rtos-project/pull/339).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: trunner/config.py [hunk 1]
```diff
@@ -30,13 +30,18 @@ def resolve_phrtos_dir() -> Path:
 PYEXPECT_TIMEOUT = 8
 
 # Available targets for test runner.
-ALL_TARGETS = ['ia32-generic', 'host-pc', 'armv7m7-imxrt106x', 'armv7m7-imxrt117x', 'armv7m4-stm32l4']
+ALL_TARGETS = ['ia32-generic',
+               'armv7a9-zynq7000-emu',
```
Comment (nalajcie) [inline trunner/config.py:34]: I'm against adding here targets not defined in `phoenix-rtos-project`. It will:
 - increase confusion regarding the test names
 - break the build functionality of the test runner
 - the third part of the target is for the `project` part of the target

I assume the binaries/images for emulator and target HW might need to be different - but if that would be the case - we need to introduce different targets in `phoenix-rtos-project` repo first.

File: sample/test/fails/test.yaml [hunk 2]
```diff
@@ -9,19 +9,29 @@ test:
         - name: fail-output-1
           exec: test-fail-output-1
           harness: test-fail-output-1.py
+          targets:
+              exclude: [armv7a9-zynq7000-emu]
```
Comment (nalajcie) [inline sample/test/fails/test.yaml:13]: If You need to exclude the new target in almost all tests **it shouldn't be put in DEFAULT_TARGETS** for now.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -110,6 +110,7 @@ jobs:
         with:
           target: ${{ matrix.target }}
 
+
```
Comment (nalajcie) [inline .github/workflows/ci.yml:151]: unrelated change in this commit

File: trunner/config.py [hunk 4]
```diff
@@ -30,7 +30,7 @@ def resolve_phrtos_dir() -> Path:
 PYEXPECT_TIMEOUT = 8
 
 # Available targets for test runner.
-ALL_TARGETS = ['ia32-generic',
+ALL_TARGETS = ['ia32-generic-emu',
```
Comment (nalajcie) [inline trunner/config.py:33]: nah, don't do it (see my reasoning below)

File: trunner/device.py [hunk 5]
```diff
@@ -12,14 +12,25 @@
...
+        ]
     )
 }
 
 
 class RunnerFactory:
     @staticmethod
     def create(target, serial, log=False):
-        if target == 'ia32-generic':
+        if target == 'ia32-generic' or target == 'armv7a9-zynq7000-emu':
```
Comment (nalajcie) [inline trunner/device.py]:
```python
if target in [ 'ia32-generic', 'armv7a9-zynq7000']:
```
or even specify `QEMU_TARGETS = [..]`

Should probably be upgraded soon to device-configs.

File: runner.py [hunk 6]
```diff
@@ -88,6 +88,9 @@ def parse_args():
     if not args.target:
         # Run on all available targets
         args.target = config.ALL_TARGETS
+    # for now there is only test runner for ia32-generic emulator
```
Comment (nalajcie) [inline runner.py:91]: should not be needed

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -110,6 +110,44 @@ jobs:
         with:
           target: ${{ matrix.target }}
 
+  test-emu:
```
Comment (nalajcie) [inline .github/workflows/ci.yml:113]: nah, don't ^C^V, don't jungle target names in CI, thus would introduce only more confusion

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -71,15 +71,15 @@ jobs:
             _boot
             rootfs-${{ matrix.target }}.tar
 
-  test-emu:
+  test-host:
     needs: build
-    name: run tests on emulators
+    name: run tests on ubuntu host
```
Comment (nalajcie) [inline .github/workflows/ci.yml:76]: coincidentally true, but not because there is ubuntu specified in the next line (but because phoenixrtos/devel:latest docker image uses ubuntu:20.04 as a base image)

**Review by nalajcie — CHANGES_REQUESTED** [review]

File: sample/test/fails/test.yaml (hunk 2 above)
Comment (damianloew) [reply sample/test/fails/test.yaml:13]: That's right, including it in psh for psh tests would be better

File: .github/workflows/ci.yml (hunk 8 above)
Comment (damianloew) [reply .github/workflows/ci.yml]: Thanks for review, this and the remaining changes above won't be necessary, if there are two separate targets in phoenix-rtos-poject. I've adjusted testing process for that, but I have to talk about it with Hubert. Now there is a test pr with changed zynq target name and this branch to run ci properly (https://github.com/phoenix-rtos/phoenix-rtos-project/pull/381).

File: trunner/config.py (hunk 1 above)
Comment (damianloew) [reply trunner/config.py]: I propose `armv7a9-zynq7000emu` and `armv7a9-zynq7000hw`.

Comment (nalajcie) [reply trunner/config.py]: nah, the TARGET value should be `${TARGET_FAMILY}-${TARGET_SUBFAMILY}-${PROJECT}` - where `TARGET_SUBFAMILY` usually should mean target CPU. I would probably go for `armv7a9-zynq7000-qemu` and `armv7a9-zynq7000-zedboard` or similar (not sure if we will have multiple qemu targets for `zynq7000` or if `zedboard` is unique enough. Please consult with @HBuczynski regarding how to approach it.

Comment (damianloew) [reply trunner/config.py]: Ok, I will consult it and then make changes.

Comment (HBuczynski) [reply trunner/config.py]: I agree with @nalajcie, `armv7a9-zynq7000-qemu` and `armv7a9-zynq7000-zedboard` is fine. There is a one target for `qemu` and the `zedboard` is a unique name for the development board from `digilent`.

**Review by jsarzy — DISMISSED** [review]
lgtm

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -21,7 +21,7 @@ jobs:
       build_result: ${{ steps.build.outcome }}
     strategy:
       matrix:
-        target: ['armv7a7-imx6ull', 'armv7a9-zynq7000', 'armv7m7-imxrt106x', 'armv7m7-imxrt117x', 'armv7m4-stm32l4x6', 'host-pc', 'ia32-generic', 'riscv64-spike', 'riscv64-virt']
+        target: ['armv7a7-imx6ull', 'armv7a9-zynq7000-zedboard', 'armv7a9-zynq7000-qemu', 'armv7m7-imxrt106x', 'armv7m7-imxrt117x', 'armv7m4-stm32l4x6', 'host-pc', 'ia32-generic', 'riscv64-spike', 'riscv64-virt']
```
Comment (nalajcie) [inline .github/workflows/ci.yml:24]: this needs to be moved to `phoenix-rtos-project` due to reusable workflows introduction

File: psh/test.yaml [hunk 10]
```diff
@@ -41,6 +43,7 @@ test:
               exclude:
                   - armv7m7-imxrt106x
                   - armv7m7-imxrt117x
+                  - armv7a9-zynq7000-qemu
```
Comment (nalajcie) [inline psh/test.yaml:46]: @HBuczynski I assume it's not yet possible to create dirs on zynq-qemu?

File: trunner/device.py [hunk 11]
```diff
@@ -1,18 +1,19 @@
 # Imports of available runners
 from .runners import HostRunner, QemuRunner, IMXRT106xRunner, IMXRT117xRunner, STM32L4Runner
 
-from .config import PHRTOS_PROJECT_DIR, DEVICE_SERIAL_USB, PHOENIXD_PORT
+from .config import PHRTOS_PROJECT_DIR, DEVICE_SERIAL_USB, PHOENIXD_PORT, QEMU_TARGETS
 
 
 QEMU_CMD = {
```
Comment (nalajcie) [inline trunner/device.py:7]: not directly related to this PR but probably should be (sooner or later) moved to `QemuRunner` class

**Review by nalajcie — COMMENTED** [review]
quite nice, but we need to move CI changes to `phoenix-rtos-project`. Left some other small remarks which might be implemented.

File: psh/test.yaml (hunk 10 above)
Comment (damianloew) [reply psh/test.yaml:46]: Creating dirs works fine (test ls uses this functionality), but there may be other problems, like those related to ownership.
Screenshot from 2022-04-25 19-01-53
:

File: trunner/device.py (hunk 11 above)
Comment (damianloew) [reply trunner/device.py:7]: Do You mean `QEMU_CMD` should be moved?

File: trunner/config.py [hunk 12]
```diff
@@ -30,13 +30,20 @@ def resolve_phrtos_dir() -> Path:
...
+               'armv7a9-zynq7000-qemu',
+               'host-pc',
+               'armv7m7-imxrt106x',
+               'armv7m7-imxrt117x',
+               'armv7m4-stm32l4']
 
 # Default targets used by parser if 'target' value is absent
 DEFAULT_TARGETS = [target for target in ALL_TARGETS
-                   if target not in ('host-pc', 'armv7m4-stm32l4')]
+                   if target not in ('host-pc', 'armv7m4-stm32l4', 'armv7a9-zynq7000-qemu')]
```
[the remark this answers is not in the document]
Comment (damianloew) [reply trunner/config.py]: Yeah, I think it can be more readable, applied.

**Comment by damianloew** [comment]
> quite nice, but we need to move CI changes to `phoenix-rtos-project`. Left some other small remarks which might be implemented.

Thanks, I've applied those changes. I will add CI related things in `phoenix-rtos-project` PR after merge. I know that branch name is not valid now, but I hope it isn't a problem.

File: psh/test.yaml (hunk 10 above)
Comment (HBuczynski) [reply psh/test.yaml:47]: I confirm that `mkdir` works fine on `zynq-qemu`.

File: psh/test.yaml [hunk 13]
```diff
@@ -1,6 +1,8 @@
 test:
-    tests:
+    targets:
+      include: [armv7a9-zynq7000-qemu]
```
[the remark this answers is not in the document]
Comment (HBuczynski) [reply psh/test.yaml:4]: Why should the `armv7a9-zynq7000-qemu` be set as experimental?

File: trunner/device.py (hunk 11 above)
Comment (nalajcie) [reply trunner/device.py:7]: It should probably internal to `QemuRunner` class (so it can be instantiated by the target directly

File: psh/test.yaml (hunk 13 above)
[the remark this answers is not in the document]
Comment (nalajcie) [reply psh/test.yaml:4]: My understanding is that most of the tests don't succeed on `zynq7000-qemu` - based on previous iteration of this PR where zynq was explicitly excluded from almost all tests. So for now we mark `zynq` as exprimental and opt for explicitly include it in tests that do pass.

File: psh/test.yaml (hunk 10 above)
Comment (nalajcie) [reply psh/test.yaml:47]: ok, so should pass but doesn't - seems like a bug? (either in the target implementation or in test) - we should fix it in instead of not running the test :).

File: trunner/device.py (hunk 11 above)
Comment (damianloew) [reply trunner/device.py]: Do You mean sth I've done now, yes?

File: psh/test.yaml (hunk 10 above)
Comment (damianloew) [reply psh/test.yaml]: Apart from disabling checking ownership in the test (not needed for now, no multi-user support), I had to make one file name shorter to enable the test for all targets.

**Review by nalajcie — APPROVED** [review]
